### PR TITLE
Prepare release 3.1 (follow-up)

### PR DESCRIPTION
**Pull request by wismill**
Follow-up of #89: I merged it too eagerly.

I would like to add also the `Script_Extensions` property in #91.

### CODE REVIEW

**Comment by wismill** [comment]
@harendra-kumar in 6e362ec772cda7efcfcd89d2c7099704625192af I changed `Unfold` to a slightly different type. It removes the need for a special handling of `\NULL` and it speeds `toUpperString` _et al._  up.

**Comment by harendra-kumar** [comment]
> @harendra-kumar in [6e362ec](https://github.com/composewell/unicode-data/commit/6e362ec772cda7efcfcd89d2c7099704625192af) I changed `Unfold` to a slightly different type. It removes the need for a special handling of `\NULL` and it speeds `toUpperString` _et al._ up.

Looks better now.

**Review by harendra-kumar — APPROVED** [review]

**Comment by wismill** [comment]
@harendra-kumar So I was wondering how our case mapping would perform on `Text`. I added a benchmark and to my surprise we are actually 20% faster than `text`! It looks too good to be true though. I tried also with the texts from `unicode-transforms` and the results are consistent. What am I missing?

```
All
  Unicode.Char.Case
    toLowerText
      text:         OK (5.08s)
        4.88 ms ±  49 μs,  16 MB allocated, 199 KB copied,  13 MB peak memory
      unicode-data: OK (1.89s)
        3.63 ms ±  67 μs, 7.0 MB allocated,  77 KB copied,  13 MB peak memory, 0.74x
    toUpperText
      text:         OK (2.46s)
        4.75 ms ±  81 μs,  16 MB allocated,  99 KB copied,  13 MB peak memory
      unicode-data: OK (3.82s)
        3.76 ms ±  48 μs, 7.1 MB allocated,  80 KB copied,  13 MB peak memory, 0.79x
    toCaseFoldText
      text:         OK (2.40s)
        4.67 ms ±  33 μs,  16 MB allocated, 104 KB copied,  13 MB peak memory
      unicode-data: OK (1.95s)
        3.75 ms ±  45 μs, 7.1 MB allocated,  83 KB copied,  13 MB peak memory, 0.80x
```

**Comment by harendra-kumar** [comment]
> And of course I checked that the resulting Text output are identical. 

That's the ultimate truth then, it is faster,  the benchmarking code looks good to me on a quick look. To explain the difference we will have to look at how text implements it. It may also be due to some additional wrapper function call in text which is not inlined? Allocations in unicode-data are half compared to text.

It may be a good idea for text to use unicode-data for case conversions, not primarily due to performance but because text will not have to maintain that code itself, it will add a dependency though. @Bodigrim

**Comment by wismill** [comment]
Note that the code is inspired from `text`.

**Comment by wismill** [comment]
I even reach 50% speedup if I keep only the characters matching `isAlphabetic`. 

> It may also be due to some additional wrapper function call in text which is not inlined? Allocations in unicode-data are half compared to text.

I think my implementation is really close to the one of `text` so it is a mystery.

I think this discussion should continue on `text` repo. It may be useful to have a package (`unicode-data-text` ??) providing drop-in replacement of some functions of `text`. No need to published it at first.

**Comment by wismill** [comment]
It turns out that `text` has seen a nice speed-up thanks to Bodigrim: 

```
All
  Unicode.Char.Case
    toLowerText
      text:         OK (3.22s)
        1.56 ms ±  11 μs, 489 KB allocated,  23 B  copied,  17 MB peak memory
      unicode-data: OK (1.33s)
        2.62 ms ±  40 μs, 6.5 MB allocated,  56 KB copied,  17 MB peak memory, 1.68x
    toUpperText
      text:         OK (0.82s)
        1.62 ms ±  24 μs, 950 KB allocated,  53 B  copied,  17 MB peak memory
      unicode-data: OK (0.68s)
        2.64 ms ±  51 μs, 6.5 MB allocated,  57 KB copied,  17 MB peak memory, 1.63x
    toCaseFoldText
      text:         OK (1.65s)
        1.59 ms ±  20 μs, 957 KB allocated,  45 B  copied,  17 MB peak memory
      unicode-data: OK (0.66s)
        2.64 ms ±  50 μs, 6.5 MB allocated,  60 KB copied,  17 MB peak memory, 1.65x
```

Note that [this optimisation](https://github.com/haskell/text/commit/8ae2888a340911e088ad9507f22464dc383c944b) does not use the fusion API.

**Comment by wismill** [comment]
I create a new benchmark _without_ fusion, based on `text`’s code. Note that you need `text >= 2.0` to build it (use `cabal.project.local`).

Last benchmark with `text-HEAD`:

```
All
  Unicode.Char.Case
    toLowerText
      text:                     OK (1.64s)
        1.61 ms ±  19 μs, 485 KB allocated,  32 B  copied,  14 MB peak memory
      unicode-data (fusion):    OK (1.46s)
        2.82 ms ±  26 μs, 7.5 MB allocated, 112 KB copied,  14 MB peak memory, 1.75x
      unicode-data (no fusion): OK (0.85s)
        1.64 ms ±  32 μs, 486 KB allocated,  35 B  copied,  14 MB peak memory, 1.01x
    toUpperText
      text:                     OK (1.76s)
        1.69 ms ±  23 μs, 957 KB allocated,  57 B  copied,  14 MB peak memory
      unicode-data (fusion):    OK (1.45s)
        2.82 ms ±  26 μs, 7.5 MB allocated, 117 KB copied,  14 MB peak memory, 1.66x
      unicode-data (no fusion): OK (1.76s)
        1.69 ms ±  17 μs, 957 KB allocated,  54 B  copied,  15 MB peak memory, 1.00x
    toCaseFoldText
      text:                     OK (1.74s)
        1.70 ms ±  16 μs, 957 KB allocated,  52 B  copied,  15 MB peak memory
      unicode-data (fusion):    OK (1.49s)
        2.91 ms ±  29 μs, 7.5 MB allocated, 125 KB copied,  15 MB peak memory, 1.71x
      unicode-data (no fusion): OK (1.79s)
        1.72 ms ±  22 μs, 957 KB allocated,  48 B  copied,  15 MB peak memory, 1.01x
```

**Comment by harendra-kumar** [comment]
Lot more complicated code to extract the last bit of performance. It's ok I guess.

I am a bit uneasy with two versions of the case mapping functions. Let me see if that can be avoided.

**Comment by wismill** [comment]
To be clear: at the moment it is only for and example of use.

**Comment by harendra-kumar** [comment]
Is this correct:

```
toLowerText = caseConvertText
    (\w -> if w - 65 <= 25 then w + 32 else w)
    C.lowerCaseMapping
```

It maps the non-alphabetic ascii chars below 90 to some other chars. What am I missing?

**Comment by harendra-kumar** [comment]
Ah, maybe it operates on an unsigned type, so the underflow makes it too large.

**Comment by harendra-kumar** [comment]
@wismill I pushed a commit, please review. I kept only one version of the case mapping APIs and used the same for String APIs as well. We can implement the other version if it is ever needed by someone.

**Comment by wismill** [comment]
I added benchmark for `toTitle`. Latest result with `text-Head`:

```
All
  Unicode.Char.Case
    toLowerText
      text:                     OK (1.58s)
        1.53 ms ±  21 μs, 485 KB allocated,  36 B  copied,  15 MB peak memory
      unicode-data (fusion):    OK (1.94s)
        1.88 ms ±  23 μs, 3.5 MB allocated, 149 B  copied,  15 MB peak memory, 1.23x
      unicode-data (no fusion): OK (1.56s)
        1.53 ms ±  14 μs, 486 KB allocated,  34 B  copied,  15 MB peak memory, 1.00x
    toUpperText
      text:                     OK (0.83s)
        1.60 ms ±  25 μs, 950 KB allocated,  71 B  copied,  15 MB peak memory
      unicode-data (fusion):    OK (1.00s)
        1.93 ms ±  24 μs, 3.5 MB allocated, 142 B  copied,  15 MB peak memory, 1.20x
      unicode-data (no fusion): OK (1.65s)
        1.61 ms ±  26 μs, 957 KB allocated,  62 B  copied,  15 MB peak memory, 1.00x
    toTitleText
      text:                     OK (0.71s)
        2.79 ms ±  43 μs, 3.8 MB allocated, 135 B  copied,  15 MB peak memory
      unicode-data (fusion):    OK (1.07s)
        2.09 ms ±  35 μs, 3.4 MB allocated, 120 B  copied,  15 MB peak memory, 0.75x
    toCaseFoldText
      text:                     OK (1.65s)
        1.59 ms ±  16 μs, 957 KB allocated,  55 B  copied,  15 MB peak memory
      unicode-data (fusion):    OK (1.01s)
        1.94 ms ±  28 μs, 3.5 MB allocated, 117 B  copied,  15 MB peak memory, 1.22x
      unicode-data (no fusion): OK (0.82s)
        1.60 ms ±  28 μs, 950 KB allocated,  57 B  copied,  15 MB peak memory, 1.00x
```

File: unicode-data/Changelog.md [hunk 1]
```diff
@@ -4,10 +4,10 @@
...
 - Added full case conversions to `Unicode.Char.Case`:
 
-  - Case folding: `caseFoldMapping`, `toCaseFoldString`.
-  - Lower case: `lowerCaseMapping`, `toLowerString`.
-  - Upper case: `upperCaseMapping`, `toUpperString`.
-  - Title case: `titleCaseMapping`, `toTitleString`.
+  - Case folding: `caseFoldMapping`, `caseFoldMapping'`, `toCaseFoldString`.
+  - Lower case: `lowerCaseMapping`, `lowerCaseMapping'`, `toLowerString`.
+  - Upper case: `upperCaseMapping`, `upperCaseMapping'`, `toUpperString`.
+  - Title case: `titleCaseMapping`, `titleCaseMapping'`, `toTitleString`.
```
Comment (harendra-kumar) [inline unicode-data/Changelog.md:10]: We can remove the alternative versions now.

Comment (wismill) [reply unicode-data/Changelog.md]: Fixed.

**Comment by wismill** [comment]
I moved the `Text` functions to its own experimental package, with the corresponding tests and benchmarks. It is not meant to be published, but can facilitate experiments.

**Comment by harendra-kumar** [comment]
toTitle comparison with text seems to be failing with 8.6.5, other CIs are getting canceled because of this failure (failfast is true):

```
1) Unicode.Text.Case.toTitle Compare with `text`
       expected: "\NULA A\NUL \SOHA A\SOH \STXA A\STX \ETXA A\ETX \EOTA A\EOT \ENQA A\ENQ \ACKA A\ACK \aA A\a \bA A\b \tA A\t \nA A\n \vA A\v \fA A\f \rA A\r \SOA A\SO \SIA A
```

**Comment by wismill** [comment]
I restricted the tests to specific versions ranges of `base` and  `text`. I think we are all set to merge it.

**Comment by wismill** [comment]
- [`unicode-data-0.3.1`](https://hackage.haskell.org/package/unicode-data-0.3.1/candidate)
- [`unicode-data-names-0.1.0`](https://hackage.haskell.org/package/unicode-data-names0.1.0) (unchanged)
- [`unicode-data-scripts-0.1.0`](https://hackage.haskell.org/package/unicode-data-scripts-0.1.0/candidate)
- [`unicode-data-security-0.1.0`](https://hackage.haskell.org/package/unicode-data-security-0.1.0/candidate)

**Comment by wismill** [comment]
I’m going to merge this to allow PR for Unicode 15.

**Comment by harendra-kumar** [comment]
Seems like we can omit the `allBlockRanges` API. We can trivially get it by enumerating the blocks and mapping blockDefinition on the resulting list. Is there some rationale to having a separate API for this? We can hold it off until a need arises. Once we release an API we have to maintain it forever, so keeping the API set as simple and concise as possible is always good.

**Comment by wismill** [comment]
Agreed. I missed it. I think the rational was to test only defined characters. See #94. Candidates updated.

**Pull request merged**